File: core/src/main/java/fr/sncf/osrd/infra_state/api/DetectionSectionState.java
package fr.sncf.osrd.infra_state.api;

import fr.sncf.osrd.infra.api.reservation.DetectionSection;
import fr.sncf.osrd.infra.api.reservation.ReservationRoute;

/** Encodes the state of the detection section */
public interface DetectionSectionState {
    enum Summary {
        FREE,
        RESERVED,
        OCCUPIED,
    }

    Summary summarize();

    /** Returns the route for which this detection state is currently reserved for */
    ReservationRoute getRoute();

    /** Returns the train which currently occupies the detection section */
    ReservationTrain getOccupyingTrain();

    /** Returns the immutable detection section object */
    DetectionSection getDetectionSection();
}

File: core/src/main/kotlin/fr/sncf/osrd/dyn_infra/api/DetectionSectionState.kt
package fr.sncf.osrd.dyn_infra.api

enum class DetectionSectionState(val reserved: Boolean, val occupied: Boolean) {
    UNRESERVED_FREE(reserved = false, occupied = false),
    RESERVED_FREE(reserved = true, occupied = false),
    UNRESERVED_OCCUPIED(reserved = false, occupied = true),
    RESERVED_OCCUPIED(reserved = true, occupied = true);

    companion object {
        @JvmStatic fun get(reserved: Boolean, occupied: Boolean): DetectionSectionState {
            return if (occupied) {
                if (reserved) RESERVED_OCCUPIED else UNRESERVED_OCCUPIED
            } else {
                if (reserved) RESERVED_FREE else UNRESERVED_FREE
            }
        }
    }
}